libbeat/publisher/queue/diskqueue/serialize.go

```diff
@@ -0,0 +1,161 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Encoding / decoding routines adapted from
+// libbeat/publisher/queue/spool/codec.go.
+
+package diskqueue
+
+import (
+	"bytes"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/outputs/codec"
+	"github.com/elastic/beats/v7/libbeat/publisher"
+	"github.com/elastic/go-structform/gotype"
+	"github.com/elastic/go-structform/json"
+)
+
+type eventEncoder struct {
+	buf    bytes.Buffer
+	folder *gotype.Iterator
+}
+
+type eventDecoder struct {
+	buf []byte
+
+	parser   *json.Parser
+	unfolder *gotype.Unfolder
+}
+
+type entry struct {
+	Timestamp int64
+	Flags     uint8
+	Meta      common.MapStr
+	Fields    common.MapStr
+}
+
+func newEventEncoder() *eventEncoder {
+	e := &eventEncoder{}
+	e.reset()
+	return e
+}
+
+func (e *eventEncoder) reset() {
+	e.folder = nil
+
+	visitor := json.NewVisitor(&e.buf)
+	folder, err := gotype.NewIterator(visitor,
+		gotype.Folders(
+			codec.MakeTimestampEncoder(),
+			codec.MakeBCTimestampEncoder(),
+		),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	e.folder = folder
+}
+
+func (e *eventEncoder) encode(event *publisher.Event) ([]byte, error) {
+	e.buf.Reset()
+
+	var flags uint8
+	// TODO: handle guaranteed send?
+	/*if (event.Flags & publisher.GuaranteedSend) == publisher.GuaranteedSend {
+		flags = flagGuaranteed
+	}*/
+
+	err := e.folder.Fold(entry{
+		Timestamp: event.Content.Timestamp.UTC().UnixNano(),
+		Flags:     flags,
+		Meta:      event.Content.Meta,
+		Fields:    event.Content.Fields,
+	})
+	if err != nil {
+		e.reset()
+		return nil, err
+	}
+
+	// Copy the encoded bytes to a new array owned by the caller.
+	bytes := e.buf.Bytes()
+	result := make([]byte, len(bytes))
+	copy(result, bytes)
+
+	return result, nil
+}
+
+func newEventDecoder() *eventDecoder {
+	d := &eventDecoder{}
+	d.reset()
+	return d
+}
+
+func (d *eventDecoder) reset() {
+	unfolder, err := gotype.NewUnfolder(nil)
+	if err != nil {
+		panic(err) // can not happen
```

Comment:
You can bet money on such "never" conditions :)

Reply:
Well: this is not expressing generic confidence that the error conditions will never arise, but rather observing that `NewUnfolder` deterministically returns `nil` when called on `nil` :-) It certainly can have an error if we request an unfolder for a different / unsupported type.